services-core/src/solution_submission/retry.rs

```diff
@@ -10,170 +12,153 @@ use futures::{
+    }
 
-        let solution_submission_future = contract
+    fn submit_solution(&self, args: &Args, gas_price: f64) -> BoxFuture<FutureOutput> {
+        self.contract
             .submit_solution(
-                batch_index,
-                solution.clone(),
-                claimed_objective_value,
+                args.batch_index,
+                args.solution.clone(),
+                args.claimed_objective_value,
                 num::f64_to_u256(gas_price),
-                nonce,
+                args.nonce,
             )
-            .map(FutureOutput::SolutionSubmission);
-        futures.push(solution_submission_future.boxed());
+            .map(FutureOutput::SolutionSubmission)
+            .boxed()
+    }
 
-        let is_last_iteration = gas_price >= effective_gas_price_cap;
-        if !is_last_iteration {
-            let timeout_future = async_sleep
-                .sleep(SINGLE_ATTEMPT_TIMEOUT)
-                .map(|()| FutureOutput::Timeout);
-            futures.push(timeout_future.boxed());
-        }
+    fn wait_interval(&self) -> BoxFuture<FutureOutput> {
+        self.async_sleep
+            .sleep(GAS_PRICE_REFRESH_INTERVAL)
+            .map(|()| FutureOutput::Timeout)
+            .boxed()
+    }
+
+    async fn retry_(&self, args: Args) -> Result<(), MethodError> {
+        let mut futures = FuturesUnordered::new();
+        let mut highest_used_gas_price = 0.0;
+        loop {
+            if let Some(gas_price) = self.new_gas_price(&args).await {
+                if gas_price >= highest_used_gas_price * MIN_GAS_PRICE_INCREASE_FACTOR {
+                    highest_used_gas_price = gas_price;
+                    futures.push(self.submit_solution(&args, gas_price));
+                }
+            }
 
-        // Like in `StableXSolutionSubmitter::submit_solution` we need to handle the situation where
-        // we observe a nonce error from one future before the completion of another. It is also
-        // possible that a previous submission transaction completes first instead of the most
-        // recent one.
-        // Unwrap because we always add the solution future above and every iteration here checks
-        // that there are still futures left.
-        while let FutureOutput::SolutionSubmission(result) = futures.next().await.unwrap() {
-            if !result.is_transaction_error() || futures.is_empty() {
-                return result;
+            futures.push(self.wait_interval());
+
+            // Like in `StableXSolutionSubmitter::submit_solution` we need to handle the situation where
+            // we observe a nonce error from one future before the completion of another. It is also
+            // possible that a previous submission transaction completes first instead of the most
+            // recent one.
+            // Unwrap because the timeout future always exists.
+            while let FutureOutput::SolutionSubmission(result) = futures.next().await.unwrap() {
```

Comment:
Thanks for pointing out the specific confusing points. I am going to add more comments to make this easier to understand (the loop, the different cases).
One thing to realize about the cancellation logic is that it is quite independent of the gas price logic. As long as the cancellation always cancels at higher gas price I don't see anything that can go wrong. But still I agree that the cancellation should be integrated more with the retry code.
For the state machine it's hard to judge without trying. I will see how it looks when integrating cancellation.
The confusion about `futures` being a vector is unfortunate but here we a have a line `let mut futures = FuturesUnordered::new()` that imo makes it clear enough and which should make the reader check out the FuturesUnordered documentation. But I think it will become more obvious anyway when I add more comments here.